### ISSUE TITLE
Python (subprocess?) gets blocked on latest master

### ISSUE THREAD

**Issue by syrusakbary** [wasmerio/wasmer#4303]
In latest `master`, executing python and running `import sys; help(sys)` 

```
$ RUST_BACKTRACE=1 target/release/wasmer run python --dir=.
Python 3.12.0 (heads/wasix-compatibility-dirty:186ebc7b8b, Oct  3 2023, 10:47:09) [Clang 15.0.6 ] on wasix
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys;
>>> help(sys)
wasm execute failed [/bin/sh] - not found
thread '<unnamed>' panicked at 'rwlock write lock would result in deadlock', /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys/unix/locks/pthread_rwlock.rs:149:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: wasmer_wasix::utils::owned_mutex_guard::write_owned
   3: wasmer_wasix::fs::WasiInodes::std_dev_get_mut
   4: wasmer_wasix::fs::WasiFs::flush::{{closure}}
   5: <tokio::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll
   6: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
   7: wasmer_wasix::state::env::WasiEnv::cleanup::{{closure}}
   8: wasmer_wasix::bin_factory::exec::<impl wasmer_wasix::bin_factory::BinFactory>::spawn::{{closure}}
   9: std::thread::local::LocalKey<T>::with
  10: futures_executor::local_pool::block_on
  11: wasmer_wasix::syscalls::wasix::proc_exec::proc_exec::{{closure}}
  12: wasmer_wasix::syscalls::wasix::proc_exec::proc_exec
  13: std::panicking::try
  14: corosensei::coroutine::on_stack::wrapper
  15: _stack_call_trampoline
  16: wasmer_vm::trap::traphandlers::on_host_stack
  17: wasmer::sys::externals::function::<impl wasmer::externals::function::HostFunction<T,(A1,A2,A3,A4),Rets,wasmer::externals::function::WithEnv> for Func>::function_callback::func_wrapper
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'tokio-runtime-worker' panicked at 'Error shouldn't be shared', lib/wasix/src/runners/wasi.rs:319:57
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: core::panicking::panic_display
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:150:5
   3: core::panicking::panic_str
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:134:5
   4: core::option::expect_failed
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/option.rs:2025:5
   5: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
   6: <D as wasmer_wasix::runtime::task_manager::VirtualTaskManagerExt>::spawn_and_block_on::{{closure}}
   7: <wasmer_wasix::runtime::task_manager::tokio::TokioTaskManager as wasmer_wasix::runtime::task_manager::VirtualTaskManager>::task_shared::{{closure}}
   8: tokio::runtime::task::core::Core<T,S>::poll
   9: tokio::runtime::task::harness::Harness<T,S>::poll
  10: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
  11: tokio::runtime::scheduler::multi_thread::worker::Context::run
  12: tokio::runtime::context::scoped::Scoped<T>::set
  13: tokio::runtime::context::runtime::enter_runtime
  14: tokio::runtime::scheduler::multi_thread::worker::run
  15: tokio::runtime::task::core::Core<T,S>::poll
  16: tokio::runtime::task::harness::Harness<T,S>::poll
  17: tokio::runtime::blocking::pool::Inner::run
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', lib/wasix/src/runtime/task_manager/mod.rs:392:47
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:117:5
   3: <wasmer_wasix::runners::wasi::WasiRunner as wasmer_wasix::runners::runner::Runner>::run_command
   4: wasmer_cli::commands::run::Run::run_wasi
   5: wasmer_cli::commands::run::Run::execute_inner
   6: wasmer_cli::commands::run::Run::execute
   7: wasmer_cli::cli::wasmer_main_inner
   8: wasmer_cli::cli::wasmer_main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

# What should happen
```
$ RUST_BACKTRACE=1 target/release/wasmer run python --dir=.
Python 3.12.0 (heads/wasix-compatibility-dirty:186ebc7b8b, Oct  3 2023, 10:47:09) [Clang 15.0.6 ] on wasix
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys;
>>> help(sys)
wasm execute failed [/bin/sh] - not found
wasm execute failed [/bin/sh] - not found
Help on built-in module sys:

NAME
    sys

MODULE REFERENCE
    https://docs.python.org/3.12/library/sys.html

    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    This module provides access to some objects used or maintained by the
    interpreter and to functions that interact strongly with the interpreter.

    Dynamic objects:

    argv -- command line arguments; argv[0] is the script pathname if known
    path -- module search path; path[0] is the script directory, else ''
    modules -- dictionary of loaded modules


>>>
```

**Comment by syrusakbary** [comment]
This was resolved when merging #4304. However, we need to add tests to verify this doesn't happen again

**Comment by john-sharratt** [comment]
This should fix the deadlock - the deadlock was always there however because we were not actually closing the files properly it was not being hit in master. This fixes both
https://github.com/wasmerio/wasmer/pull/4307
